File: src/test/java/com/example/the_machine/service/mapping/ThreadMapperTest.java
package com.example.the_machine.service.mapping;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import com.example.the_machine.domain.ThreadEntity;
import com.example.the_machine.dto.ThreadDTO;
import java.time.Instant;
import java.util.UUID;
import lombok.val;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

@SpringBootTest
class ThreadMapperTest {

  @Autowired
  private ThreadMapper threadMapper;

  @Test
  void testEntityToDto() {
    // Given
    val entity = createTestThreadEntity();

    // When
    val dto = threadMapper.toDto(entity);

    // Then
    assertNotNull(dto);
    assertEquals(entity.getId(), dto.id());
    assertEquals(entity.getTitle(), dto.title());
    assertEquals(entity.getCreatedAt(), dto.createdAt());
    assertEquals(entity.getLastActivityAt(), dto.lastActivityAt());
  }

  @Test
  void testDTOToEntity() {
    // Given
    val dto = createTestThreadDto();

    // When
    val entity = threadMapper.toEntity(dto);

    // Then
    assertNotNull(entity);
    assertEquals(dto.id(), entity.getId());
    assertEquals(dto.title(), entity.getTitle());
    assertEquals(dto.createdAt(), entity.getCreatedAt());
    assertEquals(dto.lastActivityAt(), entity.getLastActivityAt());
  }

  @Test
  void testNullHandling() {
    // Test null entity
    assertNull(threadMapper.toDto(null));

    // Test null DTO
    assertNull(threadMapper.toEntity(null));
  }

  @Test
  void testNullOptionalFields() {
    // Test entity with null title
    val entity = ThreadEntity.builder()
        .id(UUID.randomUUID())
        .title(null) // null title
        .createdAt(Instant.now())
        .lastActivityAt(Instant.now())
        .build();

    val dto = threadMapper.toDto(entity);

    assertNotNull(dto);
    assertEquals(entity.getId(), dto.id());
    assertNull(dto.title());
    assertEquals(entity.getCreatedAt(), dto.createdAt());
    assertEquals(entity.getLastActivityAt(), dto.lastActivityAt());
  }

  private ThreadEntity createTestThreadEntity() {
    return ThreadEntity.builder()
        .id(UUID.randomUUID())
        .title("Test Thread")
        // Use fixed timestamps to ensure consistent round-trip conversion
        .createdAt(Instant.parse("2024-01-01T10:00:00Z"))
        .lastActivityAt(Instant.parse("2024-01-01T11:00:00Z"))
        .build();
  }

  private ThreadDTO createTestThreadDto() {
    return new ThreadDTO(
        UUID.randomUUID(),
        "Test DTO Thread",
        Instant.now().minusSeconds(86400), // 1 day ago
        Instant.now()
    );
  }
}

File: src/test/kotlin/com/example/the_machine/service/mapping/ThreadMapperTest.kt
package com.example.the_machine.service.mapping

import com.example.the_machine.common.UuidV7
import com.example.the_machine.config.TestDataConfiguration
import com.example.the_machine.config.TestDataConfiguration.FIXED_INSTANT
import com.example.the_machine.config.TestDataConfiguration.FIXED_INSTANT_LATER
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID_2
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID_3
import com.example.the_machine.domain.ThreadEntity
import com.example.the_machine.dto.ThreadDto
import io.mockk.every
import io.mockk.mockkObject
import io.mockk.unmockkObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import

@SpringBootTest
@Import(TestDataConfiguration::class)
class ThreadMapperTest {

  @Autowired
  private lateinit var threadMapper: ThreadMapper

  @Test
  fun testEntityToDto() {
    mockkObject(UuidV7)
    try {
      // Given
      val testId = TEST_THREAD_ID
      every { UuidV7.newId() } returns testId
      val entity = createTestThreadEntity()

      // When
      val dto = threadMapper.toDto(entity)

      // Then
      assertNotNull(dto)
      assertEquals(entity.id, dto.id)
      assertEquals(entity.title, dto.title)
      assertEquals(entity.createdAt, dto.createdAt)
      assertEquals(entity.lastActivityAt, dto.lastActivityAt)
    } finally {
      unmockkObject(UuidV7)
    }
  }

  @Test
  fun testDTOToEntity() {
    mockkObject(UuidV7)
    try {
      // Given
      val testId = TEST_THREAD_ID_2
      every { UuidV7.newId() } returns testId
      val dto = createTestThreadDto()

      // When
      val entity = threadMapper.toEntity(dto)

      // Then
      assertNotNull(entity)
      assertEquals(dto.id, entity.id)
      assertEquals(dto.title, entity.title)
      assertEquals(dto.createdAt, entity.createdAt)
      assertEquals(dto.lastActivityAt, entity.lastActivityAt)
    } finally {
      unmockkObject(UuidV7)
    }
  }

  @Test
  fun testNullHandling() {
    mockkObject(UuidV7)
    try {
      // Test entity with null optional fields
      val testId = TEST_THREAD_ID_3
      every { UuidV7.newId() } returns testId
      val entityWithNulls = ThreadEntity(
        lastActivityAt = FIXED_INSTANT_LATER
      )

      val dto = threadMapper.toDto(entityWithNulls)
      assertNotNull(dto)
      // Note: Since ThreadDto.title is non-null String, mapper should handle null -> non-null conversion
      // This behavior depends on the mapper implementation

      // Test regular DTO to entity conversion (no null fields in DTO since they're non-null)
      val validDto = ThreadDto(
        id = TEST_THREAD_ID,
        title = "Test Title",
        createdAt = FIXED_INSTANT,
        lastActivityAt = FIXED_INSTANT_LATER
      )

      val entity = threadMapper.toEntity(validDto)
      assertNotNull(entity)
      assertEquals(validDto.title, entity.title)
    } finally {
      unmockkObject(UuidV7)
    }
  }

  private fun createTestThreadEntity(): ThreadEntity {
    return ThreadEntity(
      title = "Test Thread",
      lastActivityAt = FIXED_INSTANT_LATER
    )
  }

  private fun createTestThreadDto(): ThreadDto {
    return ThreadDto(
      id = TEST_THREAD_ID_2,
      title = "Test DTO Thread",
      createdAt = null,
      lastActivityAt = FIXED_INSTANT_LATER
    )
  }
}